### PR TITLE
Polkadot: Combined BondAndNominate message.

### LINKED ISSUE #1270: [Polkadot] Add BondAndNominate message support
combine those two into one `utility::batch` call

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Add combined BondAndNominate message.  Fixes #1270 .

## Testing instructions

Unit tests.  SignBondAndNominate_c7a016.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: src/Polkadot/Extrinsic.cpp [hunk 1]
```diff
@@ -91,6 +91,19 @@ Data Extrinsic::encodeCall(const Proto::SigningInput& input) {
             append(data, encodeCompact(value));
             // reward destination
             append(data, reward);
+        } else if (staking.has_bond_and_nominate()) {
+            // TODO
+            auto address = SS58Address(staking.bond_and_nominate().controller(), byte(input.network()));
+            auto value = load(staking.bond_and_nominate().value());
+            auto reward = byte(staking.bond_and_nominate().reward_destination());
+            // call index
+            append(data, getCallIndex(network, stakingBond));
```
Comment (hewigovens) [inline src/Polkadot/Extrinsic.cpp:100]: you can find the call index here: https://polkascan.io/polkadot/runtime-call/27-utility-batch

**Comment by catenocrypt** [comment]
@hewig, my understanding is that the batch call should:
- encode the first Bond call
- encode the second Nominate call
- encode a batch call, with a vector of the two encoded calls.

I was not able to find the sources how batch call is encoded.  Could you assist?
I have also not found a way in Explorer to show encoded version of calls, for reverse engineering / checking.

**Comment by catenocrypt** [comment]
Sample bond-and-nominate batch extrinsic created:

https://polkadot.subscan.io/extrinsic/0xc7a016f961dbf35d58feea22694e7d79ac77175a8cc40cb017bb5e87d56142ce

image

File: src/Polkadot/Extrinsic.cpp [hunk 2]
```diff
@@ -68,60 +70,103 @@ Data Extrinsic::encodeCall(const Proto::SigningInput& input) {
...
-            // num_slashing_spans
-            encode32LE(spans, data);
-        } else if (staking.has_nominate()) {
-            std::vector<SS58Address> addresses;
-            for (auto& n : staking.nominate().nominators()) {
-                addresses.push_back(SS58Address(n, network));
+        // reward destination
+        append(data, reward);
+    } else if (staking.has_bond_and_nominate()) {
+        // TODO
```
Comment (hewigovens) [inline src/Polkadot/Extrinsic.cpp:109]:
```suggestion
```

File: src/Polkadot/Extrinsic.cpp [hunk 3]
```diff
@@ -68,60 +70,103 @@ Data Extrinsic::encodeCall(const Proto::SigningInput& input) {
...
-            append(data, encodeAddresses(addresses));
-        } else if (staking.has_chill()) {
-            // call index
-            append(data, getCallIndex(network, stakingChill));
+            // recursive call
+            call2 = encodeStakingCall(staking2, network);
+        }
+
+        // build Utility.batch call
+        append(data, getCallIndex(network, utilityBatch));
```
Comment (hewigovens) [inline src/Polkadot/Extrinsic.cpp:136]: - Add a method to reuse encode batch call?

**Review by hewigovens — APPROVED** [review]
🚀

File: src/Polkadot/Extrinsic.cpp (hunk 3 above)
Comment (catenocrypt) [reply src/Polkadot/Extrinsic.cpp]: +1, also thought about it, done now

File: src/Polkadot/Extrinsic.cpp [hunk 4]
```diff
@@ -68,60 +70,105 @@ Data Extrinsic::encodeCall(const Proto::SigningInput& input) {
...
-        } else if (staking.has_chill()) {
-            // call index
-            append(data, getCallIndex(network, stakingChill));
+            // recursive call
+            call2 = encodeStakingCall(staking2, network);
+        }
+
+        auto calls = std::vector<Data>{call1, call2};
+        data = encodeBatchCall(calls, network);
+    } else if (staking.has_bond_extra()) {
```
Comment (vikmeup) [inline src/Polkadot/Extrinsic.cpp:142]: Any chance to use switch instead?

File: src/proto/Polkadot.proto [hunk 5]
```diff
@@ -59,11 +66,12 @@ message Staking {
...
-        Unbond unbond = 3;
-        WithdrawUnbonded withdraw_unbonded = 4;
-        Nominate nominate = 5;
-        Chill chill = 6;
+        BondAndNominate bond_and_nominate = 2;
+        BondExtra bond_extra = 3;
+        Unbond unbond = 4;
+        WithdrawUnbonded withdraw_unbonded = 5;
+        Nominate nominate = 6;
+        Chill chill = 7;
```
Comment (vikmeup) [inline src/proto/Polkadot.proto:74]: What does the chill do?

**Review by vikmeup — APPROVED** [review]

File: src/Polkadot/Extrinsic.cpp (hunk 4 above)
Comment (catenocrypt) [reply src/Polkadot/Extrinsic.cpp]: +1

**Pull request merged**